src/exla/exla_nif_util.h

```diff
@@ -89,7 +89,7 @@ namespace exla {
     void* ptr = enif_alloc_resource(resource_object<T>::type, sizeof(T));
     new(ptr) T(std::move(var));
     ERL_NIF_TERM ret = enif_make_resource(env, ptr);
-    enif_release_resource(ptr);
+    // enif_release_resource(ptr);
```

Comment:
Unfortunately, I was getting random segmentation faults, and commenting this out fixed them. The weird thing is, everything works fine if I do it in IEX, but switching to running the tests causes segmentation faults in unpredictable ways.

Reply:
That's because `iex` most likely holds those references in its history.

I think I actually know why the segfault happens! Imagine this code:

```
a = EXLA.parameter(0)
b = EXLA.parameter(1)
res = EXLA.add(a, b)
```

After we call `add`, we no longer reference `a` and `b`, so the GC may garbage collect it... although `res` still points to them on the C side of things.

Maybe the solution here is, as we call `EXLA.add`, something like this:

```elixir
def add(%Op{refs: a}, %Op{refs: b}) do
  case EXLA.NIF.add(a, b) do
    {:ok, ref} -> %Op{refs: [ref] ++ a ++ b}
    {:error, ...} ->
  end
end
```

Other options are:

1. build the tree on Elixir and we call EXLA only once at the end once the tree is built
2. track the references in C++ but I think that will be trickier